programs/drift/src/instructions/user.rs

```diff
@@ -1362,6 +1372,135 @@ pub fn handle_place_and_make_perp_order<'c: 'info, 'info>(
+    oracle_map: &mut OracleMap,
+    state: &State,
+) -> Result<()> {
+    // Authenticate the signature
+    let ix_idx = load_current_index_checked(ix_sysvar)?;
+    validate!(
+        ix_idx > 0,
+        ErrorCode::InvalidVerificationIxIndex,
+        "instruction index must be greater than 0"
+    )?;
+    let ix: Instruction = load_instruction_at_checked(ix_idx as usize - 1, ix_sysvar)?;
+    verify_ed25519_ix(
+        &ix,
+        &taker.authority.to_bytes(),
+        &taker_order_params_message.clone().try_to_vec()?,
+        &sig,
+    )?;
+
+    // Verify inputs and validate market index
+    taker_order_params_message.verify_all_same_market_indexes()?;
+    let taker_order_params = taker_order_params_message.swift_order_params;
+    let clock = &Clock::get()?;
+
+    if taker_order_params.len() == 0 {
+        msg!("No orders to place");
+        return Err(print_error!(ErrorCode::SwiftOrderSequenceError)().into());
+    }
+
+    // First order must be a taker order
+    let matching_taker_order_params = taker_order_params[0];
+    if matching_taker_order_params.order_type != OrderType::Market {
+        msg!("First order must be a taker order");
+        return Err(print_error!(ErrorCode::SwiftOrderSequenceError)().into());
+    }
+    let expected_order_id = taker_order_params_message.expected_order_id;
+
+    // Place taker order
+    let taker_order_sent = controller::orders::place_swift_perp_order(
+        expected_order_id,
+        state,
+        taker,
+        taker_key,
+        perp_market_map,
+        spot_market_map,
+        oracle_map,
+        clock,
+        matching_taker_order_params,
+        PlaceOrderOptions::default(),
+    )?;
+
+    if taker_order_sent && taker_order_params.len() > 1 {
+        let non_matching_orders: Vec<OrderParams> = taker_order_params[1..].to_vec();
+        for non_matching_order_param in non_matching_orders {
+            if non_matching_order_param.order_type != OrderType::TriggerLimit
+                && non_matching_order_param.order_type != OrderType::TriggerMarket
+            {
+                msg!("Non-matching orders must be trigger orders");
+                return Err(print_error!(ErrorCode::SwiftOrderSequenceError)().into());
+            }
+            controller::orders::place_perp_order(
+                state,
+                taker,
+                taker_key,
+                perp_market_map,
+                spot_market_map,
+                oracle_map,
+                clock,
+                non_matching_order_param,
+                PlaceOrderOptions::default(),
+            )?;
+        }
+    } else {
+        msg!(
+            "Orders not placed due to taker order id mismatch: taker account next order id {}, order params expected next order id {:?}",
+            taker.next_order_id,
+            matching_taker_order_params
+        );
+    }
+
+    Ok(())
+}
+
+pub fn handle_place_swift_taker_order<'c: 'info, 'info>(
```

Comment:
think it might make more sense to move these the instructions/keeper folder since keeper can execute this ix